source/CModuleSystem.cpp

```diff
@@ -0,0 +1,347 @@
+		int jumpAddress;
+		char magic[5];
+	} segment;
+#pragma pack(pop)
+
+	file.read((char*)&segment, sizeof(segment));
+	if (file.fail())
+	{
+		TRACE("Module '%s' file header read error", path);
+		return false;
+	}
+
+	// verify segment data
+	if (std::memcmp(segment.firstInstruction, Segment_First_Instruction, sizeof(Segment_First_Instruction)) != 0 ||
+		segment.jumpAddress >= 0 || // jump labels should be negative values
+		std::memcmp(segment.magic, Segment_Magic, sizeof(Segment_Magic)) != 0) // not a custom header
+	{
+		TRACE("Module '%s' load error. Custom segment not present", path);
+		return false;
+	}
+	segment.jumpAddress = abs(segment.jumpAddress); // turn label into actual file offset
+
+	// process custom headers
+#pragma pack(push, 1)
+	struct
+	{
+		char signature[4];
+		int size;
+	} header;
+#pragma pack(pop)
+
+	bool result = false; // no custom header found yet
+	while (file.tellg() < segment.jumpAddress)
+	{
+		file.read((char*)&header, sizeof(header));
+		if (file.fail() ||
+			file.tellg() > segment.jumpAddress) // read past the segment end
+		{ 
+			TRACE("Module '%s' load error. Invalid custom header", path);
+			return false;
+		}
+
+		auto headerEndPos = file.tellg();
+		headerEndPos += header.size;
+
+		// CLEO Module Exports
+		if (std::memcmp(header.signature, Header_Signature_Module_Exports, sizeof(Header_Signature_Module_Exports)) == 0)
+		{
+			if (headerEndPos > segment.jumpAddress)
+			{
+				TRACE("Module '%s' load error. Invalid size of exports header", path);
+				return false;
+			}
+
+			while (true)
+			{
+				ModuleExport e;
+
+				if (!e.LoadFromFile(file) ||
+					!file.good() ||
+					file.tellg() > headerEndPos)
+				{
+					if (e.name.empty())
+					{
+						TRACE("Module '%s' export load error.", path);
+					}
+					else
+					{
+						TRACE("Module's '%s' export '%s' load error.", path, e.name.c_str());
+					}
+					return false;
+				}
+
+				exports[e.name] = std::move(e); // move to container
+				result = true; // something useful loaded
+
+				if (file.tellg() == headerEndPos)
+				{
+					break; // all exports done
+				}
+			}
+		}
+		else // other unknown header type
+		{
+			file.seekg(headerEndPos, file.beg);
+			if (file.fail())
+			{
+				TRACE("Module '%s' load error. Error while skipping unknown header type", path);
+				return false;
+			}
+		}
+	}
+
+	if (!file.good())
+	{
+		TRACE("Module '%s' read error", path);
+		return false;
+	}
+
+	if (!result) // no usable elements found. No point to keeping this module
+	{
+		TRACE("Module '%s' skipped. Nothing found", path);
+		return false;
+	}
+
+	// get file size
+	file.seekg(0, file.end);
+	auto size = (size_t)file.tellg();
+	file.seekg(0, file.beg);
+
+	// store file data
+	data.resize(size);
+	file.read(data.data(), size);
+	if (file.fail())
+	{
+		return false;
+	}
+
+	return true;
+}
+
+const ScriptDataRef CModuleSystem::CModule::GetExport(const char* name)
+{
+	// TODO: normalize export name?
+
+	auto it = exports.find(name);
```

Comment:
I think export search should be case-insensitive. Currently nothing in the language is case-sensitive.

I think we can store export names in lowercase, then for each GetExport call convert name to lower-case before searching.

Reply:
I had similar thought. Done.